group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java

```diff
@@ -0,0 +1,419 @@
+
+        sortedPartitions.forEach(partition -> {
+            List<String> sortedMembersWithMatchingRack = rackInfo.getSortedMembersWithMatchingRack(partition, targetAssignment);
+
+            for (String memberId : sortedMembersWithMatchingRack) {
+                if (potentiallyUnfilledMembers.containsKey(memberId) && maybeAssignPartitionToMember(memberId, partition)) {
+                    unassignedPartitions.remove(partition);
+                    break;
+                }
+            }
+        });
+    }
+
+    /**
+     * Allocates the unassigned partitions to unfilled members in a round-robin fashion.
+     *
+     * If the rack-aware strategy is enabled, partitions are attempted to be assigned back to their current owners first.
+     * This is because pure stickiness without rack matching is not considered initially.
+     *
+     * If a partition couldn't be assigned to its current owner due to the quotas OR
+     * if the rack-aware strategy is not enabled, the partitions are allocated to the unfilled members.
+     */
+    private void unassignedPartitionsRoundRobinAssignment() {
+        Queue<String> roundRobinMembers = new LinkedList<>(potentiallyUnfilledMembers.keySet());
+
+        List<TopicIdPartition> sortedPartitionsList = unassignedPartitions.stream()
+            .sorted(Comparator.comparing(TopicIdPartition::topicId).thenComparing(TopicIdPartition::partitionId))
+            .collect(Collectors.toList());
+
+        Iterator<TopicIdPartition> partitionIterator = sortedPartitionsList.iterator();
+        while (partitionIterator.hasNext()) {
+            TopicIdPartition topicIdPartition = partitionIterator.next();
+            boolean assigned = false;
+
+            if (rackInfo.useRackStrategy && currentPartitionOwners.containsKey(topicIdPartition)) {
+                String currentOwner = currentPartitionOwners.get(topicIdPartition);
+                if (potentiallyUnfilledMembers.containsKey(currentOwner)) {
+                    assigned = maybeAssignPartitionToMember(currentOwner, topicIdPartition);
+                    if (!potentiallyUnfilledMembers.containsKey(currentOwner)) {
+                        roundRobinMembers.remove(currentOwner);
+                    }
+                }
+            }
+
+            for (int i = 0; i < roundRobinMembers.size() && !assigned; i++) {
+                String memberId = roundRobinMembers.poll();
+                if (potentiallyUnfilledMembers.containsKey(memberId)) {
+                    assigned = maybeAssignPartitionToMember(memberId, topicIdPartition);
+                }
+                // Only re-add the member to the end of the queue if it's still available for assignment.
+                if (potentiallyUnfilledMembers.containsKey(memberId)) {
+                    roundRobinMembers.add(memberId);
+                }
+            }
+
+            if (assigned) {
+                partitionIterator.remove();
```

Comment:
Why do we need to remove the partition from the `sortedPartitionsList`? It does not seem necessary to me.

Reply:
you're right, thanks for pointing it out!